File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/TwoHandWeaponTraitEnum.java
package ai.ragu.charactermanager.enumeration;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

import java.util.HashMap;
import java.util.Map;

public enum TwoHandWeaponTraitEnum implements WeaponTraitEnum {
    TWO_HAND_D8(String.format("twoHand_%s", DieEnum.D8)),
    TWO_HAND_D10(String.format("twoHand_%s", DieEnum.D10)),
    TWO_HAND_D12(String.format("twoHand_%s", DieEnum.D12));

    private final String value;
    private final static Map<String, TwoHandWeaponTraitEnum> CONSTANTS = new HashMap<>();

    static {
        for (TwoHandWeaponTraitEnum c: values()) {
            CONSTANTS.put(c.value, c);
        }
    }

    TwoHandWeaponTraitEnum(String value) {
        this.value = value;
    }

    @Override
    public String toString() {
        return this.value;
    }

    @JsonValue
    public String value() {
        return this.value;
    }

    @JsonCreator
    public static TwoHandWeaponTraitEnum fromValue(String value) {
        TwoHandWeaponTraitEnum constant = CONSTANTS.get(value);
        if (constant == null) {
            throw new IllegalArgumentException(value);
        } else {
            return constant;
        }
    }
}

File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/TwoHandWeaponTraitEnum.kt
package ai.ragu.charactermanager.enumeration

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue
import java.util.*

enum class TwoHandWeaponTraitEnum(private val value: String) : WeaponTraitEnum {
    TWO_HAND_D8(String.format("twoHand_%s", DieEnum.D8)),
    TWO_HAND_D10(String.format("twoHand_%s", DieEnum.D10)),
    TWO_HAND_D12(String.format("twoHand_%s", DieEnum.D12));

    companion object {
        private val CONSTANTS: MutableMap<String, TwoHandWeaponTraitEnum> = HashMap()

        @JsonCreator
        fun fromValue(value: String?): TwoHandWeaponTraitEnum {
            val constant = CONSTANTS[value]
            return constant ?: throw IllegalArgumentException(value)
        }

        init {
            for (c in values()) {
                CONSTANTS[c.value] = c
            }
        }
    }

    override fun toString(): String {
        return value
    }

    @JsonValue
    fun value(): String {
        return value
    }

}